File: src/core/legacy/mixins/MaxWorkDaysInWeek.js

/*
 * Mixin to compute fitness regarding maximum
 * work days in week.
 */

'use strict';

const MaxWorkDaysInWeek = (Base, props) => class extends Base {
  fitness() {
    let curr1
      , result = 0
      , employeesIter = props.employees();

    // Iterate through all employees
    while (curr1 = employeesIter.next()) {
      let curr2
        , weeksIter = props.weeks();

      // Iterate through weeks
      while (curr2 = weeksIter.next()) {
        let curr3
          , workDays = 0
          , daysIter = props.days();

        // Compute number of work days in week
        while (curr3 = daysIter.next()) {
          if (this[props.shiftIndex(curr1, curr2, curr3)]) {
            workDays += 1;
          }
        }

        // Decrease fitness only if maximum of work days in week is exceeded
        result += Math.min(0, props.opts.maxWorkDaysInWeek - workDays);
      }
    }

    if (__DEV__) {
      debugs.fitness.log(this, 'MaxWorkDaysInWeek', result);
    }

    return super.fitness() + result;
  }
}

MaxWorkDaysInWeek.requiredOpts = {
  maxWorkDaysInWeek: true
};

module.exports = MaxWorkDaysInWeek;

File: src/core/main/java/workschedule/schedule/fitnesses/MaxWorkDaysInWeek.java
package workschedule.schedule.fitnesses;

import workschedule.Chromosome;
import workschedule.schedule.Day;
import workschedule.schedule.ScheduleProperties;
import workschedule.schedule.options.MaxWorkDaysInWeekOption;

import java.util.Iterator;

public final class MaxWorkDaysInWeek extends AbstractFitness {
    private final MaxWorkDaysInWeekOption option;

    public MaxWorkDaysInWeek(final ScheduleProperties props) {
        super(props);
        option = opts.forClass(MaxWorkDaysInWeekOption.class);
    }

    @Override
    public int fitness(final Chromosome chr) {
        int result = 0;

        Iterator<Integer> employeeIterator = props.employees();

        // Iterate through all employees
        while (employeeIterator.hasNext()) {
            int employee = employeeIterator.next();

            Iterator<Integer> weekIterator = props.weeks();

            // Iterate through weeks
            while (weekIterator.hasNext()) {
                int week = weekIterator.next();

                int workDays = 0;

                Iterator<Day> dayIterator = props.days();

                // Compute number of work days in week
                while (dayIterator.hasNext()) {
                    if (chr.getParam(props.getShiftIndex(employee, week, dayIterator.next()))) {
                        workDays++;
                    }
                }

                // Decrease fitness only if maximum of work days in week is exceeded
                result += Math.min(0, option.get() - workDays);
            }
        }

        return result;
    }
}
